### PR TITLE
Enable multiple asset handling

### PR DESCRIPTION
**Pull request by elanalynn**
### WHY are these changes introduced?

Part of [conditional rendering project](https://docs.google.com/document/d/1X1NhQMxz6GnTU49rDplMd1yVmX3rsj2obIfHUFQ7WkM/edit?tab=t.0).

This PR enables the processing of a `should_render` asset, as well as other assets.

### WHAT is this pull request doing?

- Maps `should_render` field from TOML to extension config
- Bundles all `should_render` modules into one file (for both `dev` and `deploy` actions)
- Sends `buildManifest` to `shopify` in order centralize file path information in one place

### How to test your changes?

[Testing instructions](https://gist.github.com/elanalynn/bff3bda620cb5d0284d0cf892e713e90)

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes

### CODE REVIEW

File: packages/app/src/cli/services/dev/extension/bundler.ts [hunk 1]
```diff
@@ -20,6 +20,11 @@ export interface FileWatcherOptions {
   payloadStore: ExtensionsPayloadStore
 }
 
+interface Overrides {
+  contents?: string
+  outputPath?: string
+}
+
 export async function setupBundlerAndFileWatcher(options: FileWatcherOptions) {
```
Comment (elanalynn) [inline packages/app/src/cli/services/dev/extension/bundler.ts:28]: Referencing [comment in original PR](https://github.com/Shopify/cli/pull/4812/files#r1836217274) - this function will go away. Where do we move this logic? @isaacroldan

File: packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts [hunk 2]
```diff
@@ -31,35 +32,43 @@ export class ESBuildContextManager {
   setAbortSignal(signal: AbortSignal) {
     this.signal = signal
     this.signal?.addEventListener('abort', async () => {
-      const allDispose = Object.values(this.contexts).map((context) => context.dispose())
+      const allDispose = Object.values(this.contexts).map((context) => context.map((ctxt) => ctxt.dispose()))
```
Comment (isaacroldan) [inline packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts:35]: This turns into an array of arrays, not sure if `Promise.all()` will handle that correctly. Better to use a `.flat()` at the end to flatten it into a single array of Promises.

File: packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts [hunk 3]
```diff
@@ -79,11 +90,36 @@ export class ESBuildContextManager {
   }
 
   async deleteContexts(extensions: ExtensionInstance[]) {
-    const promises = extensions.map((ext) => this.contexts[ext.handle]?.dispose())
+    const promises = extensions.map((ext) => this.contexts[ext.handle]?.map((context) => context.dispose()))
     await Promise.all(promises)
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts]: Ah! nice bot, but I was faster 😆

File: packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts [hunk 4]
```diff
@@ -79,11 +90,36 @@ export class ESBuildContextManager {
...
     })
   }
+
+  private async extensionEsBuildOptions(
+    extension: ExtensionInstance,
+    overrides: Pick<BundleOptions, 'stdin'> & {outputFileName?: string},
+  ) {
+    const outputPath = overrides.outputFileName
+      ? joinPath(dirname(extension.outputPath), overrides.outputFileName)
+      : extension.getOutputPathForDirectory(this.outputPath)
```
Comment (isaacroldan) [inline packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts:107]: This is not correct, the output path for the non-main assets should be:

```ts
const mainOutputPath = extension.getOutputPathForDirectory(this.outputPath)
const assetsDirectory = dirname(mainOutputPath)
const assetOutputPath = joinPath(assetsDirectory, asset.outputFileName)
```

The way it is right now, doesn't work for Consistent-dev (it is storing each file in a different folder)

My recommendation is to remove the "overrides" concept, you should always pass options to the config:
```ts
  private async extensionEsBuildOptions(
    extension: ExtensionInstance,
    stdin: StdinOptions,
    outputFileName: string,
  )
```

**Review by vividviolet — CHANGES_REQUESTED** [review]
I think we need to simplify this. Each asset should just be its own file - there's no need to create custom stdin content

File: packages/app/src/cli/models/extensions/specification.ts [hunk 5]
```diff
@@ -50,7 +60,7 @@ export interface ExtensionSpecification<TConfiguration extends BaseConfigType =
   experience: ExtensionExperience
   dependency?: string
   graphQLType?: string
-  getBundleExtensionStdinContent?: (config: TConfiguration) => string
+  getBundleExtensionStdinContent?: (config: TConfiguration) => {main: string; assets?: Asset[]}
```
[the remark this answers is not in the document]
Comment (elanalynn) [reply packages/app/src/cli/models/extensions/specification.ts:63]: We didn't want to change the existing flow for other extensions.

[the remark this answers is not in the document]
Comment (vividviolet) [reply packages/app/src/cli/models/extensions/specification.ts:63]: But you are still changing it? It went from a string to an object

[the remark this answers is not in the document]
Comment (elanalynn) [reply packages/app/src/cli/models/extensions/specification.ts:63]: But main is still a string? Is this something we could change in a follow up?

File: packages/app/src/cli/services/build/extension.ts [hunk 6]
```diff
@@ -91,12 +91,14 @@ export async function buildUIExtension(extension: ExtensionInstance, options: Ex
     env.APP_URL = options.appURL
   }
 
+  const {main, assets} = extension.getBundleExtensionStdinContent()
```
Comment (vividviolet) [inline packages/app/src/cli/services/build/extension.ts:94]: I think we should revert this change and not touch the stdin since we no longer need to use it for other assets. We can do a simple map for the build options since we have access to the config

File: packages/app/src/cli/services/build/extension.ts [hunk 7]
```diff
@@ -106,6 +108,25 @@ export async function buildUIExtension(extension: ExtensionInstance, options: Ex
       stdout: options.stdout,
       sourceMaps: extension.isSourceMapGeneratingExtension,
     })
+    if (assets) {
```
Comment (vividviolet) [inline packages/app/src/cli/services/build/extension.ts:111]:
```suggestion
   if (extension.configuration.extension_points) {
      const promises: Promise<void>[] = []
      extension.configuration.extension_points.forEach((extensionPoint) => {
        // Start of Selection
        Object.entries(extensionPoint.build_manifest.assets).forEach(([identifier, asset]) => {
          if (identifier === AssetIdentifier.Main) {
            return
          }

          promises.push(
            bundleExtension({
              minify: true,
              outputPath: joinPath(dirname(asset.filepath), `${extension.configuration.handle}-conditions.js`),
              entryPoints: [joinPath(dirname(extension.outputPath), asset.module)],
              environment: options.environment,
              env,
            }),
          )
        })
      })
      await Promise.all(promises)
    }
```

File: packages/app/src/cli/services/build/extension.ts (hunk 6 above)
Comment (vividviolet) [reply packages/app/src/cli/services/build/extension.ts:94]: Nevermind, it looks like `extension.getBundleExtensionStdinContent` is being called in multiple places including the `createContexts` method in `app-watcher-esbuild` which isn't even using this code path. It would be nice to refactor one day but it doesn't block this from shipping.

**Review by vividviolet — APPROVED** [review]
Tophat'd

File: packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts [hunk 8]
```diff
@@ -31,35 +32,46 @@ export class ESBuildContextManager {
...
+      contexts.push(context)
+
+      if (assets) {
+        await Promise.all(
+          assets.map(async (asset) => {
+            const esbuildOptions = await this.extensionEsBuildOptions(
+              {
+                contents: asset.content,
+                resolveDir: extension.directory,
+                loader: 'ts',
```
Comment (isaacroldan) [inline packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts:65]: Is this intended to be `ts`? can it be anything else?

File: packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts [hunk 9]
```diff
@@ -31,35 +32,46 @@ export class ESBuildContextManager {
...
-        sourceMaps: true,
-      })
-
+        mainOutputPath,
+      )
       const context = await esContext(esbuildOptions)
-      this.contexts[extension.handle] = context
+      contexts.push(context)
+
+      if (assets) {
```
Comment (isaacroldan) [inline packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts:58]: I didn't love how this function turned out so I rewrote here:

```ts
  async createContexts(extensions: ExtensionInstance[]) {
    const promises = extensions.map(async (extension) => {
      const {main, assets} = extension.getBundleExtensionStdinContent()
      const mainOutputPath = extension.getOutputPathForDirectory(this.outputPath)
      const esbuildOptions = await this.extensionEsBuildOptions(
        {
          contents: main,
          resolveDir: extension.directory,
          loader: 'tsx',
        },
        mainOutputPath,
      )
      const mainContexPromise = esContext(esbuildOptions)

      const assetContextPromises = (assets ?? []).map(async (asset) => {
        const esbuildOptions = await this.extensionEsBuildOptions(
          {
            contents: asset.content,
            resolveDir: extension.directory,
            loader: 'ts',
          },
          joinPath(dirname(mainOutputPath), asset.outputFileName),
        )
        return esContext(esbuildOptions)
      })

      this.contexts[extension.handle] = await Promise.all(assetContextPromises.concat(mainContexPromise))
    })

    await Promise.all(promises)
  }
```

File: packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts (hunk 8 above)
Comment (belalsj) [reply packages/app/src/cli/services/dev/app-events/app-watcher-esbuild.ts:65]: Yes, for this use case we only need `ts`. But as a future improvement, this can be configurable from within the build_manifest object.

File: packages/app/src/cli/models/extensions/specifications/ui_extension.ts [hunk 10]
```diff
@@ -25,13 +40,31 @@ export const UIExtensionSchema = BaseSchema.extend({
...
+      const buildManifest: BuildManifest = {
+        assets: {
+          [AssetIdentifier.Main]: {
+            filepath: `dist/${config.handle}.js`,
+            module: targeting.module,
+          },
+          ...(targeting.should_render?.module
+            ? {
+                [AssetIdentifier.ShouldRender]: {
+                  filepath: `dist/${config.handle}-conditions.js`,
```
[the remark this answers is not in the document]
Comment (belalsj) [reply packages/app/src/cli/models/extensions/specifications/ui_extension.ts:52]: This is intentional. We wanted the full path to be driven by one source of truth (cli). As such, if we wanted to change that in the future, we don't have to make any changes to core.

[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/models/extensions/specifications/ui_extension.ts:52]: but this breaks the output of `app-watcher-esbuild`, you end up with the conditions file being nested in 
`/<extension-id>/dist/dist/...`

File: packages/app/src/cli/services/dev/extension/payload.ts [hunk 11]
```diff
@@ -24,6 +25,43 @@
...
+              assets,
+            }
+          }
+
+          return extensionPoint
+        },
+      ),
+    )
+
+    const extensionPoints = getExtensionPoints(extensionPointsWithAssets, url)
```
Comment (isaacroldan) [inline packages/app/src/cli/services/dev/extension/payload.ts:63]: A few proposals for this function:

- Use `UIExtensionSchemaType['extension_points']` to get the right types for the extension points
- By using the new Asset interface and a for loop, you can avoid the extra `reduce`

So the implementation is simplified quite a bit:

```ts
const extensionPointsTyped: UIExtensionSchemaType['extension_points'] = extension.configuration.extension_points
    const extensionPointsWithAssets = await Promise.all(
      extensionPointsTyped.map(async (extensionPoint) => {
        const buildManifest = extensionPoint.build_manifest
        if (!buildManifest.assets) return extensionPoint
        const assets: {[key: string]: Asset} = {}

        for (const [name, asset] of Object.entries(buildManifest.assets)) {
          assets[name] = {
            name,
            url: `${url}${joinPath('/assets/', asset.filepath)}`,
            // eslint-disable-next-line no-await-in-loop
            lastUpdated: (await fileLastUpdatedTimestamp(joinPath(dirname(extension.outputPath), asset.filepath))) ?? 0,
          }
        }

        return {...extensionPoint, assets}
      }),
    )
```

Comment (isaacroldan) [reply packages/app/src/cli/services/dev/extension/payload.ts:63]: One more thing, you might want to extract this into a proper function and call it inside `getExtensionPoints`, because `getExtensionPoints` is doing some validations on the type of the extension points, makes sense to apply this transform AFTER those validations.

File: packages/app/src/cli/services/build/extension.ts [hunk 12]
```diff
@@ -106,6 +108,25 @@ export async function buildUIExtension(extension: ExtensionInstance, options: Ex
...
+      await Promise.all(
+        assets.map(async (asset) => {
+          await bundleExtension({
+            minify: true,
+            outputPath: joinPath(dirname(extension.outputPath), `${extension.configuration.handle}-conditions.js`),
```
Comment (isaacroldan) [inline packages/app/src/cli/services/build/extension.ts:116]: In here you are not using the output path defined in the asset, this should be `asset.filepath` right?
But then you'll have the same problem of the double nested `dist/dist` folder.
